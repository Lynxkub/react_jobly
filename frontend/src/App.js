import React , {useEffect, useState} from 'react';
import './App.css';
import NavBar from './NavBar';
import Routes from './Routes';
import JoblyApi from './api';



function App() {
 
  const [userToken , setUserToken] = useState(localStorage.getItem('token'))
  const [username , setUsername] = useState(localStorage.getItem('username'))

  useEffect(() => {
    localStorage.setItem('token' , userToken)
  } , [userToken])


  JoblyApi.token = localStorage.getItem('token');
  
  const jobsAppliedTo = localStorage.getItem('jobs');
  if(!jobsAppliedTo){
    localStorage.setItem('jobs' , JSON.stringify([' ']));
  }
 

  const signUp = async (newUser) => {
  setUserToken(await JoblyApi.register(newUser));
  localStorage.setItem('token' , userToken)
  setUsername(localStorage.getItem('username'));
}

  const login = async (user) => {
    setUserToken(await JoblyApi.login(user))
    localStorage.setItem('token' , userToken)
    setUsername(localStorage.getItem('username'));
  }

  const logout =  () => {
    setUserToken(undefined);
    setUsername(localStorage.getItem('username'));
  

  }



  return (
    <div className='fill-window'>
    <NavBar token = {userToken} username = {username} />
    <Routes signUp = {signUp} login = {login} logout = {logout} token = {userToken} />
    </div>
  );
}

export default App;

