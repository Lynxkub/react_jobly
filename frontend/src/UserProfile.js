import React , {useEffect, useState} from 'react'
import JoblyApi from './api'
import './home.css'
const UserProfile = () => {

    const INITIAL_STATE = {
        username : "",
        firstName : "",
        lastName : "",
        email : "",
        password : ""
    }

    const [formData , setFormData] = useState(INITIAL_STATE)
    const userToken = localStorage.getItem('token');

    useEffect(() => {
        async function getUserData () {
            const currUser = localStorage.getItem('username')
            const userData = await JoblyApi.getUser(currUser);
            setFormData({
                username : userData.username,
                firstName : userData.firstName,
                lastName : userData.lastName,
                email : userData.email
            })
        }
        getUserData()
    } , [] )

    const handleChange = (e) => {
        const {name , value} = e.target;
        setFormData(formData => ({
            ...formData,
            [name] : value
        }))
    }

    const handleSubmit = async (e) => {
        e.preventDefault();
        console.log(JoblyApi.token);
        await JoblyApi.editProfile(formData.username ,  {firstName : formData.firstName , lastName : formData.lastName , email : formData.email , password: formData.email})
    }
    if(userToken === 'undefined') {
        return (
            <div>
                <p>You must be logged in to view this page</p>
            </div>
        )
    }
    return (
        <div className='d-flex align-items-center bg-logged-in'>
             <div className='container-md bg-background w-50 p-4'>
            <h2 className='text-center text-light'>Edit Your Profile</h2>
            <h4 className='text-center text-light'>Make changes to your First Name , Last Name , or Email</h4>
          
            <form onSubmit = {handleSubmit}>
                <div className='mb-3'>
                <label htmlFor='firstName' className='form-label text-light'>First Name</label>
                <input 
                type = 'text'
                name='firstName'
                value={formData.firstName}
                data-testid='firstName'
                onChange={handleChange}
                className='form-control'
                />
                </div>
                <div className='mb-3'>
                <label htmlFor='lastName' className='form-label text-light'>Last Name</label>
                <input 
                type = 'text'
                name='lastName'
                value={formData.lastName}
                data-testid='lastName'
                onChange={handleChange}
                className='form-control'
                />
                </div>
                <div className='mb-3'>
                <label htmlFor='email' className='form-label text-light'>Email</label>
                <input 
                type = 'email'
                name='email'
                value={formData.email}
                data-testid='email'
                onChange={handleChange}
                className='form-control'
                />
                </div>
                <div className='mb-3'>
                <label htmlFor='password' className='form-label text-light'>Password</label>
                <input 
                type = 'password'
                placeholder='Enter Password to Submit Changes'
                name='password'
                value={formData.password}
                data-testid='password'
                onChange={handleChange}
                className='form-control'
                />
                </div>
                <button className='btn btn-md btn-dark'>Submit Changes</button>
            </form>
           </div>
        </div>
    )
}


export default UserProfile;


// now to create a function to update (patch) a users information on edit page