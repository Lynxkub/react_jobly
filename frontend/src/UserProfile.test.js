import React from 'react';
import {render , fireEvent} from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import '@testing-library/jest-dom/extend-expect';
import UserProfile from './UserProfile';
import JoblyApi from './api';
import App from './App';


test('it renders without crashing' , () => {
    render(
        <MemoryRouter>
            <UserProfile />
        </MemoryRouter>
    )
})

test('it matches snapshot' , () => {
    const {asFragment} = render(
        <MemoryRouter>
            <UserProfile />
        </MemoryRouter>
    )
    expect(asFragment()).toMatchSnapshot();
})


test('it updates user profile correctly' , () => {
    const {getByText , getByTestId} = render(
        <MemoryRouter initialEntries={['/']}>
            <App />
        </MemoryRouter>
    )

    // Login to site first
    const login = getByText('Login');
    fireEvent.click(login);
    
  
    const username = getByTestId('username');
    const password = getByTestId('password');
    const signInButton = getByText('Submit')
    fireEvent.change(username , {target : {value : 'test'}});
    fireEvent.change(password , {target : {value : 'testtest'}});
    fireEvent.click(signInButton);

    setTimeout(() => {
        const firstName = getByTestId('firstName')
        const lastName = getByTestId('lastName')
        const email = getByTestId('email')
        const editPassword = getByTestId('password')
        const submit = getByText('Submit Changes');
    
        fireEvent.change(firstName , {target : {value : 'testEdit'}})
        fireEvent.change(editPassword , {target : {value : 'testtest'}})
        fireEvent.click(submit);
    } , 1000)
 

    // check database to see if it matches the change above

    setTimeout( async () => {
        const userData = await JoblyApi.getUser('test');
        expect(userData).toEqual({
            username : 'test',
            firstName : 'testEdit',
            lastName : 'test',
            email : 'test@test.com'
        })
    } , 1000)
})