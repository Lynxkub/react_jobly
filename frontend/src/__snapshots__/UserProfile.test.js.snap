// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h2>
      Edit Your Profile
    </h2>
    <h4>
      Make changes to your First Name , Last Name , or Email
    </h4>
    <form>
      <label
        for="firstName"
      >
        First Name
      </label>
      <input
        data-testid="firstName"
        name="firstName"
        type="text"
        value=""
      />
      <label
        for="lastName"
      >
        Last Name
      </label>
      <input
        data-testid="lastName"
        name="lastName"
        type="text"
        value=""
      />
      <label
        for="email"
      >
        Email
      </label>
      <input
        data-testid="email"
        name="email"
        type="email"
        value=""
      />
      <label
        for="password"
      >
        Password
      </label>
      <input
        data-testid="password"
        name="password"
        placeholder="Enter Password to Submit Changes"
        type="password"
        value=""
      />
      <button>
        Submit Changes
      </button>
    </form>
  </div>
</DocumentFragment>
`;
